lib/model/message.dart

```diff
@@ -1,21 +1,332 @@
+    };
+    if (!isStateTransitionValid) {
+      throw StateError('Unexpected state transition: $oldState -> $newState');
+    }
+
+    outboxMessage._state = newState;
+    for (final view in _messageListViews) {
+      if (oldState == OutboxMessageState.hidden) {
+        view.addOutboxMessage(outboxMessage);
+      } else {
+        view.notifyListenersIfOutboxMessagePresent(localMessageId);
+      }
+    }
+  }
+
+  /// Send a message and create an entry of [OutboxMessage].
+  Future<void> _outboxSendMessage({
+    required MessageDestination destination,
+    required String content,
+    required String? realmEmptyTopicDisplayName,
+  }) async {
+    assert(!_disposed);
+    final localMessageId = _nextLocalMessageId++;
+    assert(!outboxMessages.containsKey(localMessageId));
+
+    final now = ZulipBinding.instance.utcNow().millisecondsSinceEpoch ~/ 1000;
+    _outboxMessages[localMessageId] = switch (destination) {
+      StreamDestination(:final streamId, :final topic) => StreamOutboxMessage(
+        localMessageId: localMessageId,
+        selfUserId: selfUserId,
+        timestamp: now,
+        conversation: StreamConversation(
+          streamId,
+          topic.processLikeServer(
+            // Processing this just once on creating the outbox message
+            // allows an uncommon bug, because either of these values can change.
+            // During the outbox message's life, a topic processed from
+            // "(no topic)" could become stale/wrong when zulipFeatureLevel
+            // changes; a topic processed from "general chat" could become
+            // stale/wrong when realmEmptyTopicDisplayName changes.
+            //
+            // Shrug. The same effect is caused by an unavoidable race:
+            // an admin could change the name of "general chat"
+            // (i.e. the value of realmEmptyTopicDisplayName)
+            // concurrently with the user making the send request,
+            // so that the setting in effect by the time the request arrives
+            // is different from the setting the client last heard about.
+            zulipFeatureLevel: zulipFeatureLevel,
+            realmEmptyTopicDisplayName: realmEmptyTopicDisplayName),
+          displayRecipient: null),
+        content: content),
+      DmDestination(:final userIds) => DmOutboxMessage(
+        localMessageId: localMessageId,
+        selfUserId: selfUserId,
+        timestamp: now,
+        conversation: DmConversation(allRecipientIds: userIds),
+        content: content),
+    };
+
+    _outboxMessageDebounceTimers[localMessageId] = Timer(kLocalEchoDebounceDuration, () {
+      assert(!_disposed);
+      assert(outboxMessages.containsKey(localMessageId),
+        'The timer should have been canceled when the outbox message was removed.');
+      _outboxMessageDebounceTimers.remove(localMessageId);
+      _updateOutboxMessage(localMessageId, newState: OutboxMessageState.waiting);
+    });
```

Comment:
Let's pull this callback out to its own method:
```suggestion
    _outboxMessageDebounceTimers[localMessageId] = Timer(
      kLocalEchoDebounceDuration, () => _handleOutboxDebounce(localMessageId));
```

That will help shorten this method and make its overall structure easier to see; and it'll also make clear there's a narrow interface between this overall method and the callback, with just the one bit of data `localMessageId` passing between them (vs. all the other parameters and locals that this method otherwise would offer to the callback's closure).